clean_all.py

```diff
@@ -0,0 +1,265 @@
+        :param key: key to get instance name/id
+        :param **kwargs: keyword arguments passed to API list request
+        """
+        singular_name = name[:-1] if name.endswith("s") else name
+        logger.info(f"Deleting {singular_name}: {instance.get(key, 'unknown id')}")
+        try:
+            endpoint.delete(**kwargs).execute()
+        except Exception as err:  # pylint: disable=broad-except
+            logger.warning(f"Failed to delete {singular_name}: {err}")
+
+    def _delete_all_in_location(self, location):
+        raise NotImplementedError
+
+    def _delete_in_all_locations(self, locations, object_name: str):
+        for location in locations:
+            logger.info(f"Deleting {object_name} in: {location}")
+            try:
+                self._delete_all_in_location(location=location)
+            except HttpError as err:
+                if int(err.resp["status"]) >= 500:
+                    # Ignore server errors
+                    pass
+                elif "Unexpected location" in str(err):
+                    pass
+                else:
+                    raise
+
+
+class ComputeClient(BaseDiscoveryClient):
+    endpoint = "compute"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._locations = []
+        self._zones = []
+
+    def _delete_all_in_location(self, location):
+        raise NotImplementedError(
+            "This function is not implemented for compute endpoint as there is more than one resource"
+        )
+
+    def _refresh_locations_and_zones(self, force: bool = False):
+        if self._locations and not force:
+            return
+
+        locations_objects = self._iterate(
+            endpoint=self.client.regions(), project=self.project_id
+        )
+        locations, zones = [], []
+        for loc in locations_objects:
+            locations.append(loc["name"])
+            zones.extend([z.split("/")[-1] for z in loc.get("zones", [])])
+
+        self._zones = zones
+        self._locations = locations
+
+    @property
+    def locations(self):
+        if not self._locations:
+            self._refresh_locations_and_zones(force=True)
+        return self._locations
+
+    @property
+    def zones(self) -> List[str]:
+        if not self._zones:
+            self._refresh_locations_and_zones(force=True)
+        return self._zones
+
+    def _delete_all(self, endpoint_name: str) -> None:
+        """
+        Iterates through all zones, lists object under given ``endpoint_name``
+        and then deletes those objects.
+        """
+        for zone in self.zones:
+            logger.info(f"Deleting compute {endpoint_name} in {zone}")
+            endpoint = getattr(self.client, endpoint_name)()
+            for obj in self._iterate(
+                endpoint=endpoint, project=self.project_id, zone=zone
+            ):
+                if SKIP_LABEL not in obj.get("labels", {}) and self.is_stale(
```

Comment:
I think in case of disks I think we should also be looking if list is present in`obj` under `users` key and make sure it is empty.

Reply:
Good idea, fixed in b880473